From b2d476680d6d666e92521eec3955ab4eef609077 Mon Sep 17 00:00:00 2001
From: Cody Peter Mello <cody.mello@joyent.com>
Date: Wed, 15 Mar 2017 22:44:34 +0000
Subject: [PATCH] RELENG-750 Want CLI tool for launching builds Reviewed by:
 Trent Mick <trent.mick@joyent.com> Approved by: Trent Mick
 <trent.mick@joyent.com>

---
 tools/launch-build              | 81 +++++++++++++++++++++++++++++++++
 tools/ls-missing-release-builds |  4 +-
 2 files changed, 83 insertions(+), 2 deletions(-)
 create mode 100755 tools/launch-build

diff --git a/tools/launch-build b/tools/launch-build
new file mode 100755
index 0000000..c0d2c69
--- /dev/null
+++ b/tools/launch-build
@@ -0,0 +1,81 @@
+#! /bin/bash
+#
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+#
+
+#
+# Copyright 2017, Joyent, Inc.
+#
+
+#
+# This script allows starting Jenkins builds from the command-line. To
+# use this, you will need to get an API token for a Jenkins user, and
+# provide it to the script through the JENKINS_AUTH environment variable
+# or the -u option, in the form <user>:<api token>.
+#
+
+PROGNAME=$0
+
+function usage() {
+	if [[ -n $1 ]]; then
+		printf "%s: %s\n" "$PROGNAME" "$1" >&2
+	fi
+	printf "usage: $PROGNAME [-v] [-H url] [-b BRANCH] [-t TRY_BRANCH] [-u auth] project\n" >&2
+	exit 2
+}
+
+while getopts ':H:u:b:t:hv' opt; do
+	case $opt in
+	H) JENKINS_URL=$OPTARG;;
+	u) JENKINS_AUTH=$OPTARG;;
+	b) BRANCH=$OPTARG;;
+	t) TRY_BRANCH=$OPTARG;;
+	v) VERBOSE=true;;
+	h) usage;;
+	:) usage "-$OPTARG requires an argument";;
+	\?) usage "illegal option: -$OPTARG";;
+	esac
+done
+
+PROJECT=${@:$OPTIND:1}
+
+if [[ -z $PROJECT ]]; then
+	usage "project to build must be specified"
+fi
+
+if [[ -z $JENKINS_URL ]]; then
+	JENKINS_URL=https://jenkins.joyent.us
+fi
+
+if [[ -z $JENKINS_AUTH ]]; then
+	usage "JENKINS_AUTH must be set to <user>:<api token>"
+fi
+
+if [[ -n $BRANCH && -n $TRY_BRANCH ]]; then
+	usage "-b and -t cannot be used together"
+fi
+
+if [[ -n $BRANCH ]]; then
+	printf "Building %s with BRANCH=%s\n" "$PROJECT" "$BRANCH"
+	BUILD_PARAM=`printf '{"name":"BRANCH", "value": "%s"}' $BRANCH`
+elif [[ -n $TRY_BRANCH ]]; then
+	printf "Building %s with TRY_BRANCH=%s\n" "$PROJECT" "$BRANCH"
+	BUILD_PARAM=`printf '{"name":"TRY_BRANCH", "value": "%s"}' $TRY_BRANCH`
+fi
+
+if [[ -n $VERBOSE ]]; then
+	CURL_OPTS=(-i)
+else
+	CURL_OPTS=(-s)
+fi
+
+CRUMB_URL="$JENKINS_URL/crumbIssuer/api/xml?xpath=concat(//crumbRequestField,\":\",//crumb)"
+BUILD_URL="$JENKINS_URL/job/$PROJECT/build"
+
+# Fetch the CSRF token to send in our request's headers
+CRUMB=`curl ${CURL_OPTS[@]} --user "$JENKINS_AUTH" "$CRUMB_URL"`
+
+curl ${CURL_OPTS[@]} -X POST -H "$CRUMB" "$BUILD_URL" --user "$JENKINS_AUTH" \
+    --data-urlencode json="{\"parameter\":[$BUILD_PARAM]}"
diff --git a/tools/ls-missing-release-builds b/tools/ls-missing-release-builds
index 1c8dded..dd46018 100755
--- a/tools/ls-missing-release-builds
+++ b/tools/ls-missing-release-builds
@@ -6,7 +6,7 @@
 #
 
 #
-# Copyright (c) 2014, Joyent, Inc.
+# Copyright 2017, Joyent, Inc.
 #
 
 #
@@ -73,6 +73,6 @@ echo "$targets" | while read target public; do
     latest=$(mls $latest_mpath 2>/dev/null || true)
     if [[ -z "$latest" ]]; then
         echo "# $target, '$latest_mpath' does not exist" >&2
-        echo "open \"https://jenkins.joyent.us/job/$target/buildWithParameters?BRANCH=$release\""
+        echo "./tools/launch-build -b $release $target"
     fi
 done
-- 
2.21.0

